services/placement/placement.go

```diff
@@ -83,27 +82,43 @@ func (p placement) NumShards() int {
 
 // Validate validates a placement
 func Validate(p services.ServicePlacement) error {
-	shardCountMap := ConvertShardSliceToMap(p.Shards())
+	shardCountMap := convertShardSliceToMap(p.Shards())
 	if len(shardCountMap) != len(p.Shards()) {
 		return errDuplicatedShards
 	}
 
 	expectedTotal := len(p.Shards()) * p.ReplicaFactor()
-	actualTotal := 0
+	totalCapacity := 0
+	totalLeaving := 0
+	totalInit := 0
 	for _, instance := range p.Instances() {
-		for _, id := range instance.Shards().ShardIDs() {
-			if count, exist := shardCountMap[id]; exist {
-				shardCountMap[id] = count + 1
-				continue
+		for _, s := range instance.Shards().All() {
+			count, exist := shardCountMap[s.ID()]
+			if !exist {
+				return errUnexpectedShards
+			}
+			if s.State() == shard.Available {
+				shardCountMap[s.ID()] = count + 1
+				totalCapacity++
+			} else if s.State() == shard.Initializing {
+				totalInit++
+				shardCountMap[s.ID()] = count + 1
+				totalCapacity++
```

Comment:
If there is a copy of this shard already available, should it not count towards the `totalCapacity`?

Otherwise `expectedTotal != totalCapacity`.

e.g. imagine a case of bringing a new node online:

```
# stage 1
hostA { shardRange = 0..1023 }
hostB { shardRange = nil }

# stage 2
hostA { shardRange = 0..1023 }
hostB { shardRange = 512..1023 }

# stage 3
hostA { shardRange = 0..511 }
hostB { shardRange = 512..1023 }
```

For this case, in stage 2, the totalCapacity would be 1536 and would not equal the expectedTotal of `1024 * 1 = 1024`.